flink-yarn/src/main/java/org/apache/flink/yarn/YarnFlinkResourceManager.java

```diff
@@ -300,6 +301,8 @@ protected void shutdownApplication(ApplicationStatus finalStatus, String optiona
 		} catch (Throwable t) {
 			LOG.error("Could not cleanly shut down the Node Manager Client", t);
 		}
+
+		self().tell(decorateMessage(PoisonPill.getInstance()), self());
```

Comment:
I would directly call `getContext().system().stop(self())`, because that way we will stop immediately processing any further messages.